packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -113,51 +108,73 @@ const withInfiniteScroll = (
         showSeeMore: false,
       };
 
-      this.onItemsFetch = props.onItemsFetch;
+      this.props.dataSource.onDataChange(this.updateData);
       this.observer = new IntersectionObserver(this.handleIntersection, {
         threshold: 1,
       });
     }
 
     componentDidMount() {
-      if (this.sentinel) {
-        this.observer.observe(this.sentinel);
-      }
+      this.loadMore().then(newState => {
+        this.setState(newState);
+      });
     }
 
-    componentDidUpdate() {
+    componentDidUpdate(prevProps) {
       if (this.sentinel && this.state.index > 0) {
         this.observer.observe(this.sentinel);
       }
+      if (this.props.dataSource !== prevProps.dataSource) {
+        prevProps.dataSource.removeListener(this.updateData);
+        this.props.dataSource.onDataChange(this.updateData);
+        this.loadMore(0).then(newState => {
+          this.setState(newState);
+        });
+      }
     }
 
-    componentWillUnMount() {
+    componentWillUnmount() {
+      this.props.dataSource.removeListener(this.updateData);
       if (this.sentinel) {
         this.observer.unobserve(this.sentinel);
       }
     }
 
-    loadMore() {
-      const { index, elementsToRender } = this.state;
+    updateData = () => {
+      const { index } = this.state;
+      const { elementsPerScroll } = this.props;
+      this.props.dataSource
+        .fetchItems(index - elementsPerScroll, elementsPerScroll)
+        .then(updatedElements => {
+          this.setState({
+            elementsToRender: updatedElements,
+          });
+        });
+    };
+
+    loadMore(indexParam) {
+      const index = indexParam || this.state.index;
```

Comment:
This is not working as expected as `0` will be treated as false here and state will be used. If you test in the stories you'll see that changing stories is not reseting the datasource state. A possible solution is `indexParam != null ? indexPara : this.state.index`.

Reply:
true, changed